src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/SlidingWindowRateLimiter.cs

```diff
@@ -0,0 +1,414 @@
+            if (_requestCount >= requestCount && _requestCount != 0)
+            {
+                if (requestCount == 0)
+                {
+                    // Edge case where the check before the lock showed 0 available permits but when we got the lock some permits were now available
+                    lease = SuccessfulLease;
+                    return true;
+                }
+
+                // a. if there are no items queued we can lease
+                // b. if there are items queued but the processing order is newest first, then we can lease the incoming request since it is the newest
+                if (_queueCount == 0 || (_queueCount > 0 && _options.QueueProcessingOrder == QueueProcessingOrder.NewestFirst))
+                {
+                    _idleSince = null;
+                    _requestsPerSegment[_currentSegmentIndex] += requestCount;
+                    _requestCount -= requestCount;
+                    Debug.Assert(_requestCount >= 0);
+                    lease = SuccessfulLease;
+                    return true;
+                }
+            }
+
+            lease = null;
+            return false;
+        }
+
+        /// <summary>
+        /// Attempts to replenish request counters in a window.
+        /// </summary>
+        /// <returns>
+        /// False if <see cref="SlidingWindowRateLimiterOptions.AutoReplenishment"/> is enabled, otherwise true.
+        /// Does not reflect if permits were replenished.
+        /// </returns>
+        public override bool TryReplenish()
+        {
+            if (_options.AutoReplenishment)
+            {
+                return false;
+            }
+            Replenish(this);
+            return true;
+        }
+
+        private static void Replenish(object? state)
+        {
+            SlidingWindowRateLimiter limiter = (state as SlidingWindowRateLimiter)!;
+            Debug.Assert(limiter is not null);
+
+            // Use Stopwatch instead of DateTime.UtcNow to avoid issues on systems where the clock can change
+            long nowTicks = Stopwatch.GetTimestamp();
+            limiter!.ReplenishInternal(nowTicks);
+        }
+
+        // Used in tests that test behavior with specific time intervals
+        private void ReplenishInternal(long nowTicks)
+        {
+            // method is re-entrant (from Timer), lock to avoid multiple simultaneous replenishes
+            lock (Lock)
+            {
+                if (_disposed)
+                {
+                    return;
+                }
+
+                if ((long)((nowTicks - _lastReplenishmentTick) * TickFrequency) < ReplenishmentPeriod.Ticks)
+                {
+                    return;
+                }
+
+                _lastReplenishmentTick = nowTicks;
+
+                int oldSegmentRequestCount = _requestsPerSegment[_currentSegmentIndex];
+                _currentSegmentIndex = (_currentSegmentIndex + 1) % _options.SegmentsPerWindow;
```

Comment:
Why was this moved after grabbing the `oldSegmentRequestCount`? Shouldn't it be before so we add back the correct number of requests? If it should be moved back, can we add a test that verifies the behavior?